New multi-course structure for training materials (#486)
nextflow-io/training: docs/hello_nextflow/01_hello_world.md

### 3.1. Add a `publishDir` directive to the `sayHello` process

In the workflow script file `hello-world.nf`, make the following code modification:

_Before:_

```groovy title="hello-world.nf" linenums="6"
process sayHello {

    output:
        path 'output.txt'
```

_After:_

```groovy title="hello-world.nf" linenums="6"
process sayHello {

    publishDir 'results', mode: 'copy'

    output:
        path 'output.txt'
```

### 3.2. Run the workflow again

Now run the modified workflow script:

```bash
nextflow run hello-world.nf
```

The log output should look very familiar:

```console title="Output"
 N E X T F L O W   ~  version 24.10.0

 ┃ Launching `hello-world.nf` [mighty_lovelace] DSL2 - revision: 6654bc1327

executor >  local (1)
[10/15498d] sayHello [100%] 1 of 1 ✔
```

This time, Nextflow has created a new directory called `results/`.
Our `output.txt` file is in this directory.
If you check the contents it should match the output in the work subdirectory.
This is how we move results files outside of the working directories conveniently.

It is also possible to set the `publishDir` directive to make a symbolic link to the file instead of actually copying it.

> This is preferable when you're dealing with very large files.

Reviewer: ```suggestion
This is preferable when you're dealing with very large files you don't need to retain longer term.
```

I still maintain that the practical applications of the default symliking are few.

### 1.1. Examine the overall code structure

Let's open the `hello-world.nf` script in the editor pane.

!!! note

    The file is in the `hello-nextflow` directory, which should be your current working directory.
    You can either click on the file in the file explorer, or type `ls` in the terminal and Cmd+Click (MacOS) or Ctrl+Click (PC) on the file to open it.

```groovy title="hello-world.nf" linenums="1"
#!/usr/bin/env nextflow

/*
 * Use echo to print 'Hello World!' to a file
 */
process sayHello {

    output:
        path 'output.txt'

    script:
    """
    echo 'Hello World!' > output.txt
    """
}

workflow {

    // emit a greeting
    sayHello()
}
```

As you can see, a Nextflow script involves two main types of core components: one or more **processes**, and the **workflow** itself.
Each **process** describes what operation(s) the corresponding step in the pipeline should accomplish, while the **workflow** describes the dataflow logic that connects the various steps.

Let's take a closer look at the **process** block first, then we'll look at the **workflow** block.

> ### 1.2 The `process` definition

Reviewer 2: ```suggestion
### 1.2 The `process` block
```

The first block of code describes a **process**.
The process definition starts with the keyword `process`, followed by the process name and finally the process body delimited by curly braces.
The process body must contain a script block which specifies the command to run, which can be anything you would be able to run in a command line terminal.

Here we have a **process** called `sayHello` that writes its **output** to a file named `output.txt`.

```groovy title="hello-world.nf" linenums="3"
/*
 * Use echo to print 'Hello World!' to a file
 */
process sayHello {

    output:
        path 'output.txt'

    script:
    """
    echo 'Hello World!' > output.txt
    """
}
```

This a very minimal process definition that just contains an `output` definition and the `script` to execute.

The `output` definition includes the `path` qualifier, which tells Nextflow this should be handled as a path (includes both directory paths and files).
Another common qualifier is `val`.

!!! note

    The output definition does not _determine_ what output will be created.
    It simply _declares_ what is the expected output, so that Nextflow can look for it once execution is complete.

>     This is necessary for verifying that the command was executed successfully and for passing the output to downstream processes if needed.

Reviewer 2: ```suggestion
    This is necessary for verifying that the command was executed successfully and for passing the output to downstream processes if needed. Output produced that doesn't match what is declared in the output block will not be passed to downstream processes.
```

!!! warning

    This example is brittle because we hardcoded the output filename in two separate places (the script and the output blocks).
    If we change one but not the other, the script will break.
    Later, you'll learn how to use variables to avoid this problem.

> In a real-world pipeline, a process usually contains additional blocks such as directives, inputs, and conditional clauses, which we'll introduce later in this training course.

Reviewer 2: Will we? I did a quick search and couldn't find examples of conditional clauses in this training course.

!!! tip

    If you browse the contents of the task subdirectory in the VSCode file explorer, you'll see all the files right away.
    However, the log files are set to be invisible in the terminal, so if you want to use `ls` or `tree` to view them, you'll need to set the relevant option for displaying invisible files.

    ```bash
    tree -a work
    ```

You should see something like this, though the exact subdirectory names will be different on your system:

```console title="Directory contents"
work
└── a3
    └── 7be2fad5e71e5f49998f795677fd68
        ├── .command.begin
        ├── .command.err
        ├── .command.log
        ├── .command.out
        ├── .command.run
        ├── .command.sh
        ├── .exitcode
        └── output.txt
```

These are the helper and log files:

- **`.command.begin`**: Metadata related to the beginning of the execution of the process call
- **`.command.err`**: Error messages (`stderr`) emitted by the process call
- **`.command.log`**: Complete log output emitted by the process call
- **`.command.out`**: Regular output (`stdout`) by the process call
- **`.command.run`**: Full script run by Nextflow to execute the process call

> - **`.command.sh`**: The command that was run by the process call call

Reviewer 2: ```suggestion
- **`.command.sh`**: The command that was actually run by the process call
```

- **`.exitcode`**: The exit code resulting from the command

The `.command.sh` file is especially useful because it tells you what command Nextflow actually executed.
In this case it's very straightforward, but later in the course you'll see commands that involve some interpolation of variables.
When you're dealing with that, you need to be able to check exactly what was run, especially when troubleshooting an issue.

The actual output of the `sayHello` process is `output.txt`.
Open it and you will find the `Hello World!` greeting, which was the expected result of our minimalist workflow.

```console title="output.txt" linenums="1"
Hello World!
```

### Takeaway

You know how to decipher a simple Nextflow script, run it and find the output and relevant log files in the work directory.

### What's next?

Learn how to manage your workflow executions conveniently.

---

## 3. Manage workflow executions

Knowing how to launch workflows and retrieve outputs is great, but you'll quickly find there are a few other aspects of workflow management that will make your life easier, especially if you're developing your own workflows.

> Here we show you how to use the `publishDir` directive to TODO, the `resume` feature for when you need to re-launch the same workflow, and how to delete older work directories with `nextflow clean`.

Reviewer 2: ```suggestion
Here we show you how to use the `publishDir` directive to store in an output folder all the main results from your pipeline run, the `resume` feature for when you need to re-launch the same workflow, and how to delete older work directories with `nextflow clean`.
```

### 3.1. Publish outputs

As you have just learned, the output produced by our pipeline is buried in a working directory several layers deep.
This is done on purpose; Nextflow is in control of this directory and we are not supposed to interact with it.

However, that makes it inconvenient to retrieve outputs that we care about.

Fortunately, Nextflow provides a way to manage this more conveniently, called the `publishDir` directive, which acts at the process level.

> This directive tells Nextflow to copy the output(s) of the process to a designated output directory.

Reviewer 2: ```suggestion
This directive tells Nextflow to publish the output(s) of the process to a designated output directory. By default, the outputs are published as symbolic links from the `work` directory.
```

It allows us to retrieve the desired output file without having to dig down into the work directory.

#### 3.1.1. Add a `publishDir` directive to the `sayHello` process

In the workflow script file `hello-world.nf`, make the following code modification:

_Before:_

```groovy title="hello-world.nf" linenums="6"
process sayHello {

    output:
        path 'output.txt'
```

_After:_

```groovy title="hello-world.nf" linenums="6"
process sayHello {

    publishDir 'results', mode: 'copy'

    output:
        path 'output.txt'
```

#### 3.1.2. Run the workflow again

Now run the modified workflow script:

```bash
nextflow run hello-world.nf
```

The log output should look very familiar:

```console title="Output" linenums="1"
 N E X T F L O W   ~  version 24.10.0

Launching `hello-world.nf` [jovial_mayer] DSL2 - revision: 35bd3425e5

executor >  local (1)
[62/49a1f8] sayHello | 1 of 1 ✔
```

This time, Nextflow has created a new directory called `results/`.
Our `output.txt` file is in this directory.
If you check the contents it should match the output in the work subdirectory.

> This is how we move results files outside of the working directories conveniently.

Reviewer 2: ```suggestion
This is how we publish results files outside of the working directories conveniently.
```

> It is also possible to set the `publishDir` directive to make a symbolic link to the file instead of actually copying it.

Reviewer 2: ```suggestion
```

> This is preferable when you're dealing with very large files you don't need to retain longer term.

Reviewer 2: ```suggestion
When you're dealing with very large files that you don't need to retain for long, you may prefer to set the `publishDir` directive to make a symbolic link to the file instead of copying it.
```

> However, if you delete the work directory as part of a cleanup operation, you will lost access to the file, so always make sure you have actual copies of everything you care about before deleting anything.

Reviewer 2: ```suggestion
However, if you delete the work directory as part of a cleanup operation, you will lose access to the file, so always make sure you have actual copies of everything you care about before deleting anything.
```

!!! note

>     A newer syntax option had been proposed to make it possible to declare and publish workflow-level outputs, documented [here](https://www.nextflow.io/docs/latest/workflow.html#publishing-outputs).

Reviewer 2: ```suggestion
    A newer syntax option documented [here](https://www.nextflow.io/docs/latest/workflow.html#publishing-outputs) has been proposed to make it possible to declare and publish workflow-level outputs.
```

Nextflow has an option called `-resume` that allows you to do this.
Specifically, in this mode, any processes that have already been run with the exact same code, settings and inputs will be skipped.
This means Nextflow will only run processes that you've added or modified since the last run, or to which you're providing new settings or inputs.

There are two key advantages to doing this:

- If you're in the middle of developing your pipeline, you can iterate more rapidly since you only have to run the process(es) you're actively working on in order to test your changes.
- If you're running a pipeline in production and something goes wrong, in many cases you can fix the issue and relaunch the pipeline, and it will resume running from the point of failure, which can save you a lot of time and compute.

To use it, simply add `-resume` to your command and run it:

```bash
nextflow run hello-world.nf -resume
```

The console output should look similar.

```console title="Output" linenums="1"
 N E X T F L O W   ~  version 24.10.0

Launching `hello-world.nf` [golden_cantor] DSL2 - revision: 35bd3425e5

[62/49a1f8] sayHello | 1 of 1, cached: 1 ✔
```

> Look for the `cached:` bit that has been added in the process status line (line 6), which means that Nextflow has recognized that it has already done this work and simply re-used the result from the previous successful run.

Reviewer 2: ```suggestion
Look for the `cached:` bit that has been added in the process status line (line 5), which means that Nextflow has recognized that it has already done this work and simply re-used the result from the previous successful run.
```

You can also see that the work subdirectory hash is the same as in the previous run.
Nextflow is literally pointing you to the previous execution and saying "I already did that over there."

!!! note

    When your re-run a pipeline with `resume`, Nextflow does not overwrite any files written to a `publishDir` directory by any process call that was previously run successfully.

### 3.3. Delete older work directories

During the development process, you'll typically run your draft pipelines a large number of times, which can lead to an accumulation of very many files across many subdirectories.
Since the subdirectories are named randomly, it is difficult to tell from their names what are older vs. more recent runs.

> Nextflow includes a convenient `-clean` command that can automatically delete the work subdirectories for past runs that you no longer care about, with several [options](https://www.nextflow.io/docs/latest/reference/cli.html#clean) to control what will be deleted.

Reviewer 2: ```suggestion
Nextflow includes a convenient `clean` subcommand that can automatically delete the work subdirectories for past runs that you no longer care about, with several [options](https://www.nextflow.io/docs/latest/reference/cli.html#clean) to control what will be deleted.
```

!!! note

    The parameter name (at the workflow level) does not have to match the input variable name (at the process level).
    We're just using the same word because that's what makes sense and keeps the code readable.

In the workflow block, make the following code change:

_Before:_

```groovy title="hello-world.nf" linenums="24"
// emit a greeting
sayHello()
```

_After:_

```groovy title="hello-world.nf" linenums="24"
// emit a greeting
sayHello(params.greeting)
```

This tells Nextflow to run the `sayHello` process on the value provided through the `--greeting` parameter.

#### 4.1.4. Run the workflow command again

Let's run it!

```bash
nextflow run hello-world.nf --greeting 'Bonjour le monde!'
```

If you made all three edits correctly, you should get another successful execution:

```console title="Output" linenums="1"
 N E X T F L O W   ~  version 24.10.0

Launching `hello-world.nf` [elated_lavoisier] DSL2 - revision: 7c031b42ea

executor >  local (1)
[4b/654319] sayHello | 1 of 1 ✔
```

Be sure to open up the output file to check that you now have the new version of the greeting.

> ```console title="output.txt" linenums="1"

Reviewer 2: ```suggestion
```console title="results/output.txt" linenums="1"
```

In many cases, it makes sense to supply a default value for a given parameter so that you don't have to specify it for every run.

#### 4.2.1. Set a default value for the CLI parameter

Let's give the `greeting` parameter with a default value by declaring it before the workflow definition.

```groovy title="hello-world.nf" linenums="22"
/*
 * Pipeline parameters
 */
params.greeting = 'Holà mundo!'
```

!!! tip

    You can put the parameter declaration inside the workflow block if you prefer. Whatever you choose, try to group similar things in the same place so you don't end up with declarations all over the place.

#### 4.2.2. Run the workflow again without specifying the parameter

Now that you have a default value set, you can run the workflow again without having to specify a value in the command line.

```bash
nextflow run hello-world.nf
```

The console output should look the same.

```console title="Output" linenums="1"
 N E X T F L O W   ~  version 24.10.0

Launching `hello-world.nf` [determined_edison] DSL2 - revision: 3539118582

executor >  local (1)
[72/394147] sayHello | 1 of 1 ✔
```

Check the output in the results directory:

> ```console title="output.txt" linenums="1"

Reviewer 2: ```suggestion
```console title="results/output.txt" linenums="1"
```

Holà mundo!
```

Nextflow used the default value to name the output.

#### 4.2.3. Run the workflow again with the parameter to override the default value

If you provide the parameter on the command line, the CLI value will override the default value.

Try it out:

```bash
nextflow run hello-world.nf --greeting 'Konnichiwa!'
```

The console output should look the same.

```console title="Output" linenums="1"
 N E X T F L O W   ~  version 24.10.0

Launching `hello-world.nf` [elegant_faraday] DSL2 - revision: 3539118582

executor >  local (1)
[6f/a12a91] sayHello | 1 of 1 ✔
```

Now you will have the corresponding new output in your results directory.

> ```console title="output.txt" linenums="1"

Reviewer 2: ```suggestion
```console title="results/output.txt" linenums="1"
```
